// @solana/actions/lib/cjs/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ACTIONS_CORS_HEADERS = exports.BLINKS_QUERY_PARAM = exports.MEMO_PROGRAM_ID = exports.HTTPS_PROTOCOL = exports.SOLANA_ACTIONS_PROTOCOL_PLURAL = exports.SOLANA_ACTIONS_PROTOCOL = exports.SOLANA_PAY_PROTOCOL = void 0;
/** @internal */
exports.SOLANA_PAY_PROTOCOL = "solana:";
/** @internal */
exports.SOLANA_ACTIONS_PROTOCOL = "solana-action:";
/** @internal */
exports.SOLANA_ACTIONS_PROTOCOL_PLURAL = "solana-actions:";
/** @internal */
exports.HTTPS_PROTOCOL = "https:";
/** Program Id for the SPL Memo program */
exports.MEMO_PROGRAM_ID = "MemoSq4gqABAXKb96qnH8TysNcWxMyWCqXgDLGmfcHr";
/**  */
exports.BLINKS_QUERY_PARAM = "action";
/**
 * Standard headers
 */
exports.ACTIONS_CORS_HEADERS = {
    "Access-Control-Allow-Origin": "*",
    "Access-Control-Allow-Methods": "GET,POST,PUT,OPTIONS",
    "Access-Control-Allow-Headers": "Content-Type, Authorization, Content-Encoding, Accept-Encoding",
    "Content-Type": "application/json",
};
